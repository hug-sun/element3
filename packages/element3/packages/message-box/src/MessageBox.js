// #todo
import { isVNode } from 'vue'
import { createComponent } from '../../../src/use/component'
import msgboxVue from './MessageBox.vue'

let currentMsg, instance
let msgQueue = []

const defaultCallback = (action) => {
  if (currentMsg) {
    if (currentMsg.resolve) {
      if (action === 'confirm') {
        if (instance.proxy.showInput) {
          currentMsg.resolve({
            value: instance.proxy.inputValue,
            action
          })
        } else {
          currentMsg.resolve({ action })
        }
      } else if (
        currentMsg.reject &&
        (action === 'cancel' || action === 'close')
      ) {
        currentMsg.reject({ action })
      }
    }
  }
}

const initInstance = (currentMsg, VNode = null) => {
  instance = createComponent(msgboxVue, currentMsg.options, VNode)
  MessageBox.instance = instance
}

const showNextMsg = () => {
  if (msgQueue.length > 0) {
    currentMsg = msgQueue.shift()
    const options = currentMsg.options

    if (options.callback === undefined) {
      options.callback = defaultCallback
    }

    const oldCb = options.callback
    options.callback = (action, instance) => {
      oldCb(action, instance)
    }

    initInstance(currentMsg, isVNode(options.message) ? options.message : null)
    document.body.appendChild(instance.vnode.el)
  }
}

const MessageBox = function (options) {
  let callback = null
  if (options.callback) {
    callback = options.callback
  }

  if (typeof Promise !== 'undefined') {
    let promiseInstance = new Promise((resolve, reject) => {
      // eslint-disable-line
      msgQueue.push({
        options: options,
        callback: callback,
        resolve: resolve,
        reject: reject
      })
      showNextMsg()
    })
    promiseInstance.instance = instance
    return promiseInstance
  } else {
    msgQueue.push({
      options: options,
      callback: callback
    })

    showNextMsg()
    return { instance }
  }
}

MessageBox.alert = (message, title, options) => {
  if (typeof title === 'object') {
    options = title
    title = ''
  } else if (title === undefined) {
    title = ''
  }
  if (typeof message === 'object') {
    options = message
    message = ''
  }
  return MessageBox(
    Object.assign(
      {
        type: '',
        title: title,
        message: message,
        category: 'alert',
        confirmButtonText: '确认',
        cancelButtonText: '取消'
      },
      options
    )
  )
}

MessageBox.confirm = (message, title, options) => {
  if (typeof title === 'object') {
    options = title
    title = ''
  } else if (title === undefined) {
    title = ''
  }
  if (typeof message === 'object') {
    options = message
    message = ''
  }
  return MessageBox(
    Object.assign(
      {
        type: 'info',
        title: title,
        message: message,
        category: 'confirm',
        showCancelButton: true
      },
      options
    )
  )
}

MessageBox.prompt = (message, title, options) => {
  if (typeof title === 'object') {
    options = title
    title = ''
  } else if (title === undefined) {
    title = ''
  }
  if (typeof message === 'object') {
    options = message
  }
  return MessageBox(
    Object.assign(
      {
        type: '',
        title: title,
        message: message,
        showCancelButton: true,
        showInput: true,
        category: 'prompt',
        inputErrorMessage: '输入的数据不合法!'
      },
      options
    )
  )
}

MessageBox.close = () => {
  instance.proxy.closeHandle()
  msgQueue = []
  currentMsg = null
}

export default MessageBox
export { MessageBox }
