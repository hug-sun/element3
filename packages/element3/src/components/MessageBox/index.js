export { MessageBox } from './src/MessageBox.js'
