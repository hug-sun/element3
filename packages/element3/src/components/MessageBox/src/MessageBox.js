import { isVNode } from 'vue'
import { isObject, isUndefined } from '../../../utils/types'
import { createComponent } from '../../../composables/component'
import msgboxVue from './MessageBox.vue'

let currentMsg, instance
let msgQueue = []

const defaultCallback = (action) => {
  if (currentMsg && currentMsg.resolve) {
    const isConfirm = action === 'confirm'
    const isCancelOrClose = action === 'cancel' || action === 'close'
    const isReject = currentMsg.reject && isCancelOrClose
    if (isReject) {
      currentMsg.reject({ action })
      return
    }
    const isShow = instance.proxy.showInput
    const value = instance.proxy.inputValue
    const result = isShow ? { value, action } : { action }
    if (isConfirm) {
      currentMsg.resolve(result)
    }
  }
}

const initInstance = (currentMsg, VNode = null) => {
  instance = createComponent(msgboxVue, currentMsg.options, VNode)
  MessageBox.instance = instance
}

const showNextMsg = () => {
  if (msgQueue.length <= 0) return
  currentMsg = msgQueue.shift()
  const options = currentMsg.options

  if (isUndefined(options.callback)) options.callback = defaultCallback

  const oldCb = options.callback
  options.callback = (action, instance) => {
    oldCb(action, instance)
  }

  initInstance(
    currentMsg,
    isVNode(options.message) ? () => options.message : null
  )
  document.body.appendChild(instance.vnode.el)
}

const MessageBox = function (options) {
  let callback = null
  if (options.callback) {
    callback = options.callback
  }
  let promiseInstance = new Promise((resolve, reject) => {
    // eslint-disable-line
    msgQueue.push({
      options: options,
      callback: callback,
      resolve: resolve,
      reject: reject
    })
    showNextMsg()
  })
  promiseInstance.instance = instance
  return promiseInstance
}

const mergeCondition = (message, title, options) => {
  if (isObject(title)) {
    options = title
    title = ''
  } else if (isUndefined(title)) {
    title = ''
  }
  if (isObject(message)) {
    options = message
    message = ''
  }
  return Object.assign(
    {
      title: title,
      message: message,
      confirmButtonText: '确认',
      cancelButtonText: '取消'
    },
    options
  )
}

const kindOfMessageBox = {
  alert: {
    category: 'alert',
    closeOnPressEscape: false
  },
  confirm: {
    type: 'info',
    category: 'confirm',
    showCancelButton: true
  },
  prompt: {
    showInput: true,
    category: 'prompt',
    showCancelButton: true,
    inputErrorMessage: '输入的数据不合法!'
  },
  msgbox: MessageBox
}

for (let key in kindOfMessageBox) {
  MessageBox[key] = (message, title, options) => {
    return MessageBox(
      Object.assign(
        kindOfMessageBox[key],
        mergeCondition(message, title, options)
      )
    )
  }
}

MessageBox.close = () => {
  instance.proxy.closeHandle()
  msgQueue = []
  currentMsg = null
}

export default MessageBox
export { MessageBox }
