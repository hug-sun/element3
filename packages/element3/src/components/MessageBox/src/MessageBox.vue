<template>
  <transition name="msgbox-fade">
    <div
      v-show="visible"
      class="el-message-box__wrapper"
      tabindex="-1"
      @click.self="handleWrapperClick"
    >
      <div
        class="el-message-box"
        :class="[customClass, center && 'el-message-box--center']"
      >
        <div class="el-message-box__header" v-if="title !== null">
          <div class="el-message-box__title">
            <div
              :class="['el-message-box__status', icon]"
              v-if="icon && center"
            ></div>
            <span>{{ title }}</span>
          </div>
          <button
            type="button"
            class="el-message-box__headerbtn"
            aria-label="Close"
            v-if="showClose"
            @click="
              handleAction(distinguishCancelAndClose ? 'close' : 'cancel')
            "
            @keydown.enter="
              handleAction(distinguishCancelAndClose ? 'close' : 'cancel')
            "
          >
            <i class="el-message-box__close el-icon-close"></i>
          </button>
        </div>
        <div class="el-message-box__content">
          <div class="el-message-box__container">
            <div
              :class="['el-message-box__status', icon]"
              v-if="icon && !center && message !== ''"
            ></div>
            <div class="el-message-box__message" v-if="message !== ''">
              <slot>
                <p v-if="!dangerouslyUseHTMLString">
                  {{ message }}
                </p>

                <p v-else v-html="message"></p>
              </slot>
            </div>
          </div>
          <div class="el-message-box__input" v-show="showInput">
            <el-input
              @keydown.enter="handleInputEnter"
              :type="inputType"
              v-model="inputValue"
              :placeholder="inputPlaceholder"
              ref="input"
            ></el-input>
            <div
              :style="{
                visibility: !!editorErrorMessage ? 'visible' : 'hidden'
              }"
              class="el-message-box__errormsg"
            >
              {{ editorErrorMessage }}
            </div>
          </div>
        </div>
        <div class="el-message-box__btns">
          <el-button
            :loading="cancelButtonLoading"
            size="small"
            :round="roundButton"
            @click="handleAction('cancel')"
            @keydown.enter="handleAction('cancel')"
            :class="[cancelButtonClass]"
            v-show="showCancelButton"
          >
            {{ cancelButtonText }}
          </el-button>
          <el-button
            :loading="confirmButtonLoading"
            size="small"
            ref="confirm"
            :round="roundButton"
            @click="handleAction('confirm')"
            @keydown.enter="handleAction('confirm')"
            :class="['el-button--primary', confirmButtonClass]"
            v-show="showConfirmButton"
          >
            {{ confirmButtonText }}
          </el-button>
        </div>
      </div>
    </div>
  </transition>
</template>

<script type="text/babel">
import { defineComponent, reactive, toRefs, getCurrentInstance, ref } from 'vue'
import propsObject from './props.js'

import { ElInput } from '../../Input'
import { ElButton } from '../../Button'
import validateFunction from './validate'
import { useHandleList, watchElement, classIcon } from './use'
export default defineComponent({
  props: propsObject,
  components: {
    ElInput,
    ElButton
  },
  setup(props) {
    const confirmButtonLoading = ref(false)
    const state = reactive({
      action: null,
      visible: true,
      ...props
    })
    const instance = getCurrentInstance()
    const { iconClass, type } = toRefs(state)
    const { validate, editorErrorMessage } = validateFunction(state, instance)
    const {
      closeHandle,
      handleAction,
      handleInputEnter,
      handleWrapperClick
    } = useHandleList(state, instance, validate)
    const icon = classIcon(iconClass, type)
    watchElement(state, handleAction, closeHandle)
    return {
      ...toRefs(state),
      editorErrorMessage,
      icon,
      handleAction,
      handleInputEnter,
      handleWrapperClick,
      closeHandle,
      confirmButtonLoading
    }
  }
})
</script>
