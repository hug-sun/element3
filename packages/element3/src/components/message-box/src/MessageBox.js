import { isVNode } from 'vue'
import { createComponent } from '../../../use/component'
import msgboxVue from './MessageBox.vue'

let currentMsg, instance
let msgQueue = []

const defaultCallback = (action) => {
  if (currentMsg) {
    if (currentMsg.resolve) {
      if (action === 'confirm') {
        if (instance.proxy.showInput) {
          currentMsg.resolve({
            value: instance.proxy.inputValue,
            action
          })
        } else {
          currentMsg.resolve({ action })
        }
      } else if (
        currentMsg.reject &&
        (action === 'cancel' || action === 'close')
      ) {
        currentMsg.reject({ action })
      }
    }
  }
}

const initInstance = (currentMsg, VNode = null) => {
  instance = createComponent(msgboxVue, currentMsg.options, VNode)
  MessageBox.instance = instance
}

const showNextMsg = () => {
  if (msgQueue.length > 0) {
    currentMsg = msgQueue.shift()
    const options = currentMsg.options

    if (options.callback === undefined) {
      options.callback = defaultCallback
    }

    const oldCb = options.callback
    options.callback = (action, instance) => {
      oldCb(action, instance)
    }

    initInstance(currentMsg, isVNode(options.message) ? options.message : null)
    document.body.appendChild(instance.vnode.el)
  }
}

const MessageBox = function (options) {
  let callback = null
  if (options.callback) {
    callback = options.callback
  }
  let promiseInstance = new Promise((resolve, reject) => {
    // eslint-disable-line
    msgQueue.push({
      options: options,
      callback: callback,
      resolve: resolve,
      reject: reject
    })
    showNextMsg()
  })
  promiseInstance.instance = instance
  return promiseInstance
}

const MergeCondition = (message, title, options) => {
  if (typeof title === 'object') {
    options = title
    title = ''
  } else if (title === undefined) {
    title = ''
  }
  if (typeof message === 'object') {
    options = message
    message = ''
  }
  return Object.assign(
    {
      title: title,
      message: message,
      confirmButtonText: '确认',
      cancelButtonText: '取消'
    },
    options
  )
}

MessageBox.alert = (message, title, options) => {
  const defaultVal = {
    type: null,
    category: 'alert',
    closeOnPressEscape: false
  }
  return MessageBox(
    Object.assign(defaultVal, MergeCondition(message, title, options))
  )
}

MessageBox.confirm = (message, title, options) => {
  const defaultVal = {
    type: 'info',
    category: 'confirm'
  }
  return MessageBox(
    Object.assign(defaultVal, MergeCondition(message, title, options))
  )
}

MessageBox.prompt = (message, title, options) => {
  const defaultVal = {
    type: null,
    showInput: true,
    category: 'prompt',
    inputErrorMessage: '输入的数据不合法!'
  }
  return MessageBox(
    Object.assign(defaultVal, MergeCondition(message, title, options))
  )
}

MessageBox.close = () => {
  instance.proxy.closeHandle()
  msgQueue = []
  currentMsg = null
}

export default MessageBox
export { MessageBox }
