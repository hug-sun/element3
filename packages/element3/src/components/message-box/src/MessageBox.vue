<template>
  <transition name="msgbox-fade">
    <div
      v-show="visible"
      class="el-message-box__wrapper"
      tabindex="-1"
      role="dialog"
      aria-modal="true"
      @click.self="handleWrapperClick"
      :aria-label="title || 'dialog'"
    >
      <div
        class="el-message-box"
        :class="[customClass, center && 'el-message-box--center']"
      >
        <div class="el-message-box__header" v-if="title !== null">
          <div class="el-message-box__title">
            <div
              :class="['el-message-box__status', icon]"
              v-if="icon && center"
            ></div>
            <span>{{ title }}</span>
          </div>
          <button
            type="button"
            class="el-message-box__headerbtn"
            aria-label="Close"
            v-if="showClose"
            @click="
              handleAction(distinguishCancelAndClose ? 'close' : 'cancel')
            "
            @keydown.enter="
              handleAction(distinguishCancelAndClose ? 'close' : 'cancel')
            "
          >
            <i class="el-message-box__close el-icon-close"></i>
          </button>
        </div>
        <div class="el-message-box__content">
          <div class="el-message-box__container">
            <div
              :class="['el-message-box__status', icon]"
              v-if="icon && !center && message !== ''"
            ></div>
            <div class="el-message-box__message" v-if="message !== ''">
              <slot>
                <p v-if="!dangerouslyUseHTMLString">
                  {{ message }}
                </p>

                <p v-else v-html="message"></p>
              </slot>
            </div>
          </div>
          <div class="el-message-box__input" v-show="showInput">
            <el-input
              @keydown.enter="handleInputEnter"
              :type="inputType"
              v-model="inputValue"
              :placeholder="inputPlaceholder"
              ref="input"
            ></el-input>
            <div
              :style="{
                visibility: !!editorErrorMessage ? 'visible' : 'hidden'
              }"
              class="el-message-box__errormsg"
            >
              {{ editorErrorMessage }}
            </div>
          </div>
        </div>
        <div class="el-message-box__btns">
          <el-button
            :loading="cancelButtonLoading"
            size="small"
            :round="roundButton"
            @click="handleAction('cancel')"
            @keydown.enter="handleAction('cancel')"
            :class="[cancelButtonClass]"
            v-show="showCancelButton"
          >
            {{ cancelButtonText }}
          </el-button>
          <el-button
            :loading="confirmButtonLoading"
            size="small"
            ref="confirm"
            :round="roundButton"
            @click="handleAction('confirm')"
            @keydown.enter="handleAction('confirm')"
            :class="['el-button--primary', confirmButtonClass]"
            v-show="showConfirmButton"
          >
            {{ confirmButtonText }}
          </el-button>
        </div>
      </div>
    </div>
  </transition>
</template>

<script type="text/babel">
import {
  defineComponent,
  reactive,
  toRefs,
  getCurrentInstance,
  nextTick,
  unref,
  computed,
  onMounted,
  onUnmounted,
  ref,
  watch
} from 'vue'
import propsObject from './props.js'
import { isFunction } from '@vue/shared'
import { usePopup } from '../../../use/popup'
import { ElInput } from '../../Input'
import { ElButton } from '../../Button'
import { addClass, removeClass } from '../../../utils/dom'

export default defineComponent({
  props: propsObject,
  components: {
    ElInput,
    ElButton
  },
  setup(props) {
    const state = reactive({
      action: null,
      visible: true,
      ...props
    })
    const instance = getCurrentInstance()
    const { iconClass, type } = toRefs(state)
    const { validate, editorErrorMessage } = validateFunction(state, instance)
    const {
      closeHandle,
      handleAction,
      handleInputEnter,
      handleWrapperClick
    } = handleList(state, instance, validate)
    const confirmButtonLoading = ref(false)
    const icon = computed(() => {
      return unref(iconClass) || (unref(type) ? `el-icon-${unref(type)}` : '')
    })
    return {
      ...toRefs(state),
      editorErrorMessage,
      icon,
      handleAction,
      handleInputEnter,
      handleWrapperClick,
      closeHandle,
      confirmButtonLoading
    }
    function handleList(state, instance, validate) {
      const { open, close } = usePopup(state)
      const {
        visible,
        callback,
        category,
        inputType,
        closeOnHashChange,
        closeOnPressEscape,
        distinguishCancelAndClose,
        closeOnClickModal,
        action,
        beforeClose
      } = toRefs(state)
      const closeHandle = () => {
        visible.value = false
        close()
        nextTick(() => {
          unref(callback)(state.action, instance.proxy)
        })
      }
      const handleAction = (_action) => {
        if (
          unref(category) === 'prompt' &&
          _action === 'confirm' &&
          !validate()
        ) {
          return
        }
        action.value = _action
        if (isFunction(beforeClose.value)) {
          beforeClose.value(_action, instance.proxy, closeHandle)
        } else {
          closeHandle()
        }
      }
      const handleKeyup = (element = {}) => {
        if (element.code !== 'Escape') return
        if (unref(closeOnPressEscape)) {
          handleAction(unref(distinguishCancelAndClose) ? 'close' : 'cancel')
        }
      }
      const handleInputEnter = () => {
        if (unref(inputType) !== 'textarea') {
          return handleAction('confirm')
        }
      }

      const handleWrapperClick = () => {
        if (unref(closeOnClickModal)) {
          handleAction(unref(distinguishCancelAndClose) ? 'close' : 'cancel')
        }
      }
      onMounted(() => {
        if (unref(closeOnHashChange)) {
          window.addEventListener('hashchange', closeHandle)
        }
        window.addEventListener('keyup', handleKeyup)
        nextTick(() => {
          open()
        })
      })
      onUnmounted(() => {
        if (unref(closeOnHashChange)) {
          window.removeEventListener('hashchange', closeHandle)
        }
        window.removeEventListener('keyup', handleKeyup)
      })
      return {
        closeHandle,
        handleAction,
        handleWrapperClick,
        handleInputEnter
      }
    }
    function validateFunction(state, instance) {
      const editorErrorMessage = ref(null)
      const {
        category,
        inputPattern,
        inputValue,
        inputValidator,
        inputErrorMessage
      } = toRefs(state)
      const getInputElement = () => {
        const inputRefs = instance.refs.input.$refs
        return inputRefs.input || inputRefs.textarea
      }

      const validate = () => {
        if (unref(category) === 'prompt') {
          if (
            unref(inputPattern) &&
            !unref(inputPattern).test(unref(inputValue))
          ) {
            editorErrorMessage.value = unref(inputErrorMessage)

            addClass(getInputElement(), 'invalid')
            return false
          }
          const _inputValidator = unref(inputValidator)
          if (typeof _inputValidator === 'function') {
            const validateResult = _inputValidator(unref(inputValue))
            if (validateResult === false) {
              editorErrorMessage.value = unref(inputErrorMessage)
              addClass(getInputElement(), 'invalid')
              return false
            }
            if (typeof validateResult === 'string') {
              editorErrorMessage.value = validateResult
              addClass(getInputElement(), 'invalid')
              return false
            }
          }
        }
        editorErrorMessage.value = ''
        removeClass(getInputElement(), 'invalid')
        return true
      }
      watch(inputValue, (val) => {
        nextTick(() => {
          if (unref(category) === 'prompt' && val !== null) {
            validate()
          }
        })
      })
      return {
        validate,
        editorErrorMessage
      }
    }
  }
})
</script>
