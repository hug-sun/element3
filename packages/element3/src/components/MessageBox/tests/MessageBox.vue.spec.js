import { mount } from '@vue/test-utils'
import { nextTick, h, ref } from 'vue'
import MessageBox from '../src/MessageBox.vue'
describe('MessageBox.vue', () => {
  describe('snapshot', () => {
    const wrapper = mount(MessageBox)
    expect(wrapper.element).toMatchSnapshot()
  })
  describe('proprety', () => {
    it('proprety title', () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: 'chushihua'
        }
      })
      expect(wrapper.get('.el-message-box__title')).toHaveTextContent(
        'chushihua'
      )
    })
    it('proprety center', () => {
      const wrapper = mount(MessageBox, {
        props: {
          center: true
        }
      })
      expect(wrapper.get('.el-message-box--center').exists()).toBeTruthy()
    })
    it('proprety customClass', () => {
      const wrapper = mount(MessageBox, {
        props: {
          customClass: 'customClass'
        }
      })
      expect(wrapper.get('.customClass').exists()).toBeTruthy()
    })
    it('proprety iconClass', () => {
      const wrapper = mount(MessageBox, {
        props: {
          iconClass: 'iconClass'
        }
      })
      expect(wrapper.vm.icon).toBe('iconClass')
    })
    it('iconclass with center', () => {
      const wrapper = mount(MessageBox, {
        props: {
          iconClass: 'iconClass',
          center: true,
          title: 'title'
        }
      })
      expect(wrapper.get('.iconClass').exists()).toBeTruthy()
    })
    it('proprety type', () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: 'title',
          center: true,
          type: 'info'
        }
      })
      expect(wrapper.find('.el-icon-info').exists()).toBeTruthy()
    })
    it('showClose', () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: '12',
          showClose: false
        }
      })
      expect(wrapper.find('.el-message-box__headerbtn').exists()).toBeFalsy()
    })
    it('showClose toBeTruthy', () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: '12'
        }
      })
      expect(wrapper.get('.el-message-box__headerbtn').exists()).toBeTruthy()
    })
    it('proprety beforeClose', async () => {
      let object = {}
      const wrapper = mount(MessageBox, {
        props: {
          title: '12',
          beforeClose: (action, instance, done) => {
            object.action = action
            object.instance = instance
            object.done = done
          }
        }
      })
      await wrapper.get('.el-message-box__headerbtn').trigger('click')
      expect(wrapper.componentVM).toEqual(object.instance)
      expect(object.action).toBe('cancel')
    })
    it('review messageBox when value was done', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: '12',
          beforeClose: (action, instance, done) => {
            done()
          }
        }
      })
      await wrapper.get('.el-message-box__headerbtn').trigger('click')
      expect(wrapper.get('.el-message-box__headerbtn').isVisible()).toBeFalsy()
      expect(wrapper.find('.v-modal').exists()).toBeFalsy()
    })
    it('showClose lockScroll', () => {
      document.body.classList.remove('el-popup-parent--hidden')
      mount(MessageBox, {
        props: {
          title: '12',
          lockScroll: false
        },
        attachTo: document.getElementById('body')
      })
      expect(document.body.className).not.toBe('el-popup-parent--hidden')
    })
    it('meesage of normal', () => {
      const wrapper = mount(MessageBox, {
        props: {
          message: '333'
        }
      })
      expect(wrapper.get('.el-message-box__message > p')).toHaveTextContent(
        '333'
      )
    })
    it('meesage is VNode', () => {
      const v = h('p', null, [
        h('span', null, '内容可以是 '),
        h('i', { style: 'color: teal' }, 'VNode')
      ])
      const wrapper = mount(MessageBox, {
        slots: {
          default: v
        }
      })
      expect(wrapper.componentVM.$slots.default()[0]).toEqual(v)
    })
    it('dangerouslyUseHTMLString', () => {
      const wrapper = mount(MessageBox, {
        props: {
          dangerouslyUseHTMLString: true,
          message: '<div class="mmm">444</div>'
        }
      })
      expect(wrapper.get('.mmm').exists()).toBeTruthy()
    })
    it('handleInputEnter', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          inputType: 'texg'
        }
      })
      wrapper.componentVM.handleInputEnter()
      await nextTick()
      expect(wrapper.componentVM.visible).toBeFalsy()
    })
    it('closeOnPressEscape', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          closeOnPressEscape: true
        }
      })
      const event = new KeyboardEvent('keyup', {
        code: 'Escape'
      })
      window.dispatchEvent(event)
      expect(wrapper.get('.el-message-box__wrapper').isVisible()).toBeTruthy()
      expect(wrapper.componentVM.visible).toBeFalsy()
    })
    it('handleWrapperClick', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          closeOnClickModal: true
        }
      })
      await wrapper.get('.el-message-box__wrapper').trigger('click')
      await nextTick()
      expect(wrapper.componentVM.action).toBe('cancel')
    })
    it('closeOnHashChange', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          closeOnHashChange: true
        }
      })
      window.dispatchEvent(new Event('hashchange'))
      expect(wrapper.get('.el-message-box__wrapper').isVisible()).toBeTruthy()
      expect(wrapper.componentVM.visible).toBeFalsy()
    })
    it('showInput', () => {
      const wrapper = mount(MessageBox, {
        props: {
          showInput: true
        }
      })
      expect(wrapper.get('.el-message-box__input').isVisible()).toBeTruthy()
    })
    it('inputValue inputPlaceholder', () => {
      const wrapper = mount(MessageBox, {
        props: {
          showInput: true,
          inputValue: '444',
          inputPlaceholder: '555'
        }
      })
      expect(wrapper.get('.el-input__inner').element.value).toEqual('444')
      expect(wrapper.get('.el-input__inner').attributes('placeholder')).toEqual(
        '555'
      )
    })
    it('cancelButtonText, showCancelButton', () => {
      const wrapper = mount(MessageBox, {
        props: {
          cancelButtonText: 'cancel',
          showCancelButton: true,
          cancelButtonClass: 'vvv'
        }
      })
      expect(wrapper.get('.vvv')).toHaveTextContent('cancel')
      expect(wrapper.get('.vvv').isVisible()).toBeTruthy()
    })
    it('confirmButtonClass showConfirmButton, confirmButtonText', () => {
      const wrapper = mount(MessageBox, {
        props: {
          confirmButtonText: 'cancel',
          showConfirmButton: true,
          confirmButtonClass: 'mmm'
        }
      })
      expect(wrapper.get('.el-button--primary').classes()).toContain('mmm')
      expect(wrapper.get('.mmm')).toHaveTextContent('cancel')
      expect(wrapper.get('.el-message-box__btns').isVisible()).toBeTruthy()
    })
    it('proprety callback', async () => {
      let object = ref(null)
      const wrapper = mount(MessageBox, {
        props: {
          title: '12',
          callback(action) {
            object.value = action
          }
        }
      })
      await wrapper.get('.el-message-box__headerbtn').trigger('click')
      await nextTick()
      expect(object.value).toBe('cancel')
    })
    it('validate correct', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: '12',
          category: 'prompt',
          inputPattern: /[\w!#$%&'*+/=?^_`{|}~-]+(?:\.[\w!#$%&'*+/=?^_`{|}~-]+)*@(?:[\w](?:[\w-]*[\w])?\.)+[\w](?:[\w-]*[\w])?/,
          confirmButtonText: '确定',
          cancelButtonText: '取消'
        }
      })
      await wrapper.findComponent({ ref: 'input' }).setValue('2323')
      await nextTick()
      expect(wrapper.componentVM.editorErrorMessage).toBe('输入的数据不合法!')
      expect(wrapper.get('.el-input__inner').classes()).toHaveLength(2)
    })
    it('validate fail', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: '12',
          category: 'prompt',
          inputValidator() {
            return '失败'
          },
          confirmButtonText: '确定',
          cancelButtonText: '取消'
        }
      })
      await wrapper.findComponent({ ref: 'input' }).setValue('2323')
      await nextTick()
      expect(wrapper.get('.el-input__inner').classes()).toHaveLength(2)
      expect(wrapper.componentVM.editorErrorMessage).toBe('失败')
    })
    it('inputValidator', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          title: '12',
          category: 'prompt',
          inputValidator() {
            return false
          },
          confirmButtonText: '确定',
          cancelButtonText: '取消'
        }
      })
      await wrapper.findComponent({ ref: 'input' }).setValue('2323')
      await nextTick()
      expect(wrapper.get('.el-input__inner').classes()).toHaveLength(2)
    })
  })
  describe('test modal closeOnClickModal', () => {
    it('open modal', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          modal: true,
          closeOnClickModal: true
        }
      })
      await nextTick()
      expect(wrapper.get('.v-modal').exists()).toBeTruthy()
      await wrapper.get('.v-modal').trigger('click')
      await nextTick()
      expect(wrapper.find('.v-moda').exists()).toBeFalsy()
    })
    it('open modal', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          modal: true,
          closeOnClickModal: true
        }
      })
      await nextTick()
      expect(wrapper.get('.v-modal').exists()).toBeTruthy()
      await wrapper.get('.v-modal').trigger('keyup', { keyCode: 'Escape' })
      await nextTick()
      expect(wrapper.find('.v-moda').exists()).toBeFalsy()
    })
    it('open modal', async () => {
      const wrapper = mount(MessageBox, {
        props: {
          modal: true,
          closeOnClickModal: true
        }
      })
      await nextTick()
      expect(wrapper.get('.v-modal').exists()).toBeTruthy()
      await wrapper.get('.v-modal').trigger('hashchange')
      await nextTick()
      expect(wrapper.find('.v-moda').exists()).toBeFalsy()
    })
  })
})
