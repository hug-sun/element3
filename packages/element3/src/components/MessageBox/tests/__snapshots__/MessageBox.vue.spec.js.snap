// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<transition-stub>
  <div
    class="el-message-box__wrapper"
    tabindex="-1"
  >
    <div
      class="el-message-box"
    >
      <!--v-if-->
      <div
        class="el-message-box__content"
      >
        <div
          class="el-message-box__container"
        >
          <!--v-if-->
          <div
            class="el-message-box__message"
          >
            
            <p>
              {}
            </p>
            
          </div>
        </div>
        <div
          class="el-message-box__input"
          style="display: none;"
        >
          <div
            class="el-input"
          >
            
            <!--v-if-->
            <input
              class="el-input__inner"
              placeholder=""
              type="text"
            />
            <!--v-if-->
            <!--v-if-->
            <!-- 后置元素 -->
            <!--v-if-->
            
          </div>
          <div
            class="el-message-box__errormsg"
            style="visibility: hidden;"
          />
        </div>
      </div>
      <div
        class="el-message-box__btns"
      >
        <button
          class="el-button el-button--small"
          style="display: none;"
          type="button"
        >
          <!--v-if-->
          <span>
            
            确定
            
          </span>
        </button>
        <button
          class="el-button el-button--small el-button--primary"
          type="button"
        >
          <!--v-if-->
          <span>
            
            取消
            
          </span>
        </button>
      </div>
    </div>
  </div>
</transition-stub>
`;
